consensus/core/src/block_manager.rs

```diff
@@ -1,53 +1,397 @@
+
+        let mut block_manager = BlockManager::new(context.clone(), dag_state);
+
+        // create a DAG of 2 rounds
+        let all_blocks = dag(context, 2);
+
+        // WHEN
+        let accepted_blocks = block_manager
+            .accept_blocks(all_blocks.clone())
+            .expect("No error was expected");
+
+        // THEN
+        assert!(accepted_blocks.len() == 8);
+        assert_eq!(
+            accepted_blocks,
+            all_blocks
+                .iter()
+                .filter(|block| block.round() > 0)
+                .cloned()
+                .collect::<Vec<VerifiedBlock>>()
+        );
+
+        // WHEN trying to accept same blocks again, then none will be returned as those have been already accepted
+        let accepted_blocks = block_manager
+            .accept_blocks(all_blocks)
+            .expect("No error was expected");
+        assert!(accepted_blocks.is_empty());
+    }
+
+    #[test]
+    fn accept_blocks_unsuspend_children_blocks() {
+        // GIVEN
+        let (context, _key_pairs) = Context::new_for_test(4);
+        let context = Arc::new(context);
+
+        // create a DAG of 3 rounds
+        let all_blocks = dag(context.clone(), 3);
+        // keep only the non-genesis blocks
+        let mut all_blocks = all_blocks
+            .into_iter()
+            .filter(|block| block.round() > 0)
+            .collect::<Vec<_>>();
+
+        // Now randomize the sequence of sending the blocks to block manager. In the end all the blocks should be uniquely
+        // suspended and no missing blocks should exist.
+        for seed in 0..100u8 {
+            all_blocks.shuffle(&mut StdRng::from_seed([seed; 32]));
+
+            let store = Arc::new(MemStore::new());
+            let dag_state = Arc::new(RwLock::new(DagState::new(context.clone(), store.clone())));
+
+            let mut block_manager = BlockManager::new(context.clone(), dag_state);
+
+            // WHEN
+            let mut all_accepted_blocks = vec![];
+            for block in &all_blocks {
+                let accepted_blocks = block_manager
+                    .accept_blocks(vec![block.clone()])
+                    .expect("No error was expected");
+
+                all_accepted_blocks.extend(accepted_blocks);
+            }
+
+            // THEN
+            all_accepted_blocks.sort_by_key(|b| b.reference());
+            all_blocks.sort_by_key(|b| b.reference());
+
+            assert_eq!(
+                all_accepted_blocks, all_blocks,
+                "Failed acceptance sequence for seed {}",
+                seed
+            );
+            assert!(block_manager.missing_blocks().is_empty());
+            assert!(block_manager.suspended_blocks().is_empty());
+        }
+    }
+
+    /// Creates all the blocks to produce a fully connected DAG from round 0 up to `end_round`.
+    /// Note: this method also returns the genesis blocks.
```

Comment:
Let's exclude genesis blocks from BlockManager input. I don't think genesis blocks will or should be allowed to be sent over the network.

Reply:
I've added this check in order to make the implementation complete under any circumstances. I am happy to remove it as long as we indeed have those strong guarantees.

Reply:
Actually by moving genesis to DagState we get that for free anyways.